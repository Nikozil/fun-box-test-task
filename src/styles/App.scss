@import url('https://fonts.googleapis.com/css2?family=Nunito&family=Roboto&display=swap');

.app {
  display: grid;
  grid-template-areas:
    'head head'
    'interface map';
  grid-template-rows: 60px 1fr;
  grid-template-columns: 300px 1fr;

  min-height: 100vh;
  min-width: 350px;

  font-family: 'Nunito', sans-serif;

  @media screen and (max-width: 767px) {
    grid-template-columns: 200px 1fr;
  }
}
.interface {
  grid-area: interface;
  background-color: #fff;
  padding: 10px;
}

.map {
  grid-area: map;
  background-color: #f7f7f7;
}
