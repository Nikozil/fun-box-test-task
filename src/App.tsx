import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import DotsListComponent from './components/DotsListComponent';
import Header from './components/Header';
import MapComponent from './components/MapComponent';
import NewDotForm from './components/NewDotForm';
import { reorder } from './lib/Functions';
import './styles/App.scss';

function App() {
  const [center, setCenter] = useState<Coordinates>([55.734108, 37.623915]);
  const [dotsList, setDotsList] = useState<Dot[]>([]);

  const handleSubmit = (name: DotName) => {
    const newDot = {
      id: uuidv4(),
      name: name,
      coordinates: center,
    };

    setDotsList((prevlist) => [...prevlist, newDot]);
  };

  const deleteDot = (id: DotId) => {
    setDotsList((prevlist) => prevlist.filter((dot: Dot) => dot.id !== id));
  };

  const permutationDots = (startIndex: number, endIndex: number) => {
    const newList = reorder(dotsList, startIndex, endIndex);
    setDotsList(newList);
  };

  return (
    <div className="app">
      <Header />
      <div className="interface">
        <NewDotForm handleSubmit={handleSubmit} />
        <DotsListComponent
          dotsList={dotsList}
          deleteDot={deleteDot}
          permutationDots={permutationDots}
        />
      </div>
      <div className="map">
        <MapComponent
          center={center}
          setCenter={setCenter}
          dotsList={dotsList}
          setDotsList={setDotsList}
        />
      </div>
    </div>
  );
}

export default App;

export interface Dot {
  id: DotId;
  name: DotName;
  coordinates: Coordinates;
}
export type DotId = string;
type DotName = string;
export type Coordinates = number[];
