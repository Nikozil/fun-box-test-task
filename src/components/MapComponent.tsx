import React, { Fragment } from 'react';
import { GeoObject, Map, Placemark, YMaps, YMapsApi } from 'react-yandex-maps';
import { Coordinates, Dot } from '../App';
import styles from '../styles/MapComponent.module.scss';
const MapComponent: React.FC<Props> = ({
  center,
  setCenter,
  dotsList,
  setDotsList,
}) => {
  const onBoundsChangeHandler = (e: YMapsApi) => {
    setCenter(e.get('target').getCenter());
  };

  const onDragHandler = (e: YMapsApi) => {
    const newCoords = e.get('target').geometry.getCoordinates() as Coordinates;
    const id = e.get('target').properties.get('id');

    const changeCoords = (prevList: Dot[]) => {
      return prevList.map((dot: Dot) => {
        if (dot.id === id) return { ...dot, coordinates: newCoords };
        return dot;
      });
    };
    setDotsList(changeCoords);
  };

  const balloonContent = (dot: Dot) => `
                    <div class=${styles['baloon-content']}>
                   <div class=${styles['baloon-content__body']}>${dot.name}</div>
                  </div>
                    `;
  return (
    <YMaps>
      <Map
        className={styles.map}
        defaultState={{
          center: center,
          zoom: 17,
        }}
        options={{
          suppressMapOpenBlock: true,
        }}
        onBoundsChange={onBoundsChangeHandler}>
        {dotsList.map((dot, i, list) => (
          <Fragment key={dot.id}>
            <Placemark
              geometry={dot.coordinates}
              properties={{
                id: dot.id,
                balloonContent: balloonContent(dot),
              }}
              options={{
                preset: 'islands#nightCircleDotIcon',
                iconColor: '#20232a',
                cursor: 'pointer',
                draggable: true,
                hasBalloon: true,
                openBalloonOnClick: true,
                openEmptyBalloon: true,
                balloonCloseButton: false,
                hideIconOnBalloonOpen: false,
                balloonOffset: [0, -12],
              }}
              modules={['geoObject.addon.balloon']}
              onDrag={onDragHandler}
            />
            {i === 0 ? null : (
              <GeoObject
                geometry={{
                  type: 'LineString',
                  coordinates: [list[i - 1].coordinates, dot.coordinates],
                }}
                options={{
                  geodesic: true,
                  strokeWidth: 5,
                  strokeColor: '#61dafb',
                }}
              />
            )}
          </Fragment>
        ))}
      </Map>
    </YMaps>
  );
};

export default MapComponent;

interface Props {
  center: Coordinates;
  setCenter: (center: Coordinates) => void;
  dotsList: Dot[];
  setDotsList: (list: Dot[] | ChangePrevState<Dot[]>) => void;
}
type ChangePrevState<T> = (prev: T) => T;
