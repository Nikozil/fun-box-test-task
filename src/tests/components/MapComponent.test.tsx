import { render } from '@testing-library/react';
import React from 'react';
import { Dot } from '../../App';
import MapComponent from '../../components/MapComponent';

const dotsList = [
  { name: 'Dot 1', id: '213', coordinates: [10, 20] },
  { name: 'Dot 2', id: '213243', coordinates: [15, 20] },
  { name: 'Dot 3', id: '212343', coordinates: [17, 20] },
] as Dot[];
const setDotsList = jest.fn();
const setCenter = jest.fn();

describe('MapComponent tests', () => {
  test('renders correct', () => {
    const { asFragment } = render(
      <MapComponent
        center={[10, 20]}
        setCenter={setCenter}
        dotsList={dotsList}
        setDotsList={setDotsList}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
