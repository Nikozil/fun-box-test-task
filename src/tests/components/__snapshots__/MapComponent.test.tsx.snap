// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapComponent tests renders correct 1`] = `<DocumentFragment />`;
